From 97258861c85a63c2603bfa916e2bab5844824c9a Mon Sep 17 00:00:00 2001
From: Marian Chereji <marian.chereji@freescale.com>
Date: Mon, 29 Oct 2012 21:38:17 +0000
Subject: [PATCH 354/518] dpa_classifier: Enable "Remove Ethernet/IEEE 802.3 &
 VLANs" header manipulation

Enabled the Ethernet/IEEE 802.3 header remove header manipulation.

Signed-off-by: Marian Chereji <marian.chereji@freescale.com>
[Grabbed from the branch, LINUX_IR5.2.0, of
https://git.freescale.com/git-private/cgit.cgi/ppc/alu-b4860/linux.git.]
Signed-off-by: Tiejun Chen <tiejun.chen@windriver.com>
---
 drivers/staging/fsl_dpa_offload/dpa_classifier.c |   24 +++++++++++++--------
 1 files changed, 15 insertions(+), 9 deletions(-)

diff --git a/drivers/staging/fsl_dpa_offload/dpa_classifier.c b/drivers/staging/fsl_dpa_offload/dpa_classifier.c
index 70f4c83b..ffddfd6 100644
--- a/drivers/staging/fsl_dpa_offload/dpa_classifier.c
+++ b/drivers/staging/fsl_dpa_offload/dpa_classifier.c
@@ -3261,11 +3261,6 @@ static int remove_hm_check_params(const struct dpa_cls_hm_remove_params
 	BUG_ON(!remove_params);
 
 	switch (remove_params->type) {
-	case DPA_CLS_HM_REMOVE_ETHERNET:
-		pr_err("ERROR: %s, %s (%d): Unsupported HM: remove Ethernet.\n",
-			__FILE__, __func__, __LINE__);
-		return -ENOSYS;
-		break;
 	case DPA_CLS_HM_REMOVE_PPPoE:
 		pr_err("ERROR: %s, %s (%d): Unsupported HM: remove PPPoE.\n",
 			__FILE__, __func__, __LINE__);
@@ -4582,17 +4577,28 @@ static int remove_hm_update_params(struct dpa_cls_hm *premove_hm)
 
 	hm_node->params.type			= e_FM_PCD_MANIP_HDR;
 	hm_node->params.u.hdr.rmv		= TRUE;
-	hm_node->params.u.hdr.rmvParams.type	= e_FM_PCD_MANIP_RMV_GENERIC;
 	hm_node->params.u.hdr.dontParseAfterManip = TRUE;
 
 	switch (premove_hm->remove_params.type) {
+	case DPA_CLS_HM_REMOVE_ETHERNET:
+		hm_node->params.u.hdr.rmvParams.type =
+					e_FM_PCD_MANIP_RMV_BY_HDR;
+		hm_node->params.u.hdr.rmvParams.u.byHdr.type =
+					e_FM_PCD_MANIP_RMV_BY_HDR_SPECIFIC_L2;
+		hm_node->params.u.hdr.rmvParams.u.byHdr.u.specificL2 =
+					e_FM_PCD_MANIP_HDR_RMV_ETHERNET;
+		break;
 	case DPA_CLS_HM_REMOVE_PPP:
+		hm_node->params.u.hdr.rmvParams.type =
+						e_FM_PCD_MANIP_RMV_GENERIC;
 		hm_node->params.u.hdr.rmvParams.u.generic.offset =
-							PPP_HEADER_OFFSET;
+						PPP_HEADER_OFFSET;
 		hm_node->params.u.hdr.rmvParams.u.generic.size =
-							PPP_HEADER_SIZE;
+						PPP_HEADER_SIZE;
 		break;
 	case DPA_CLS_HM_REMOVE_CUSTOM:
+		hm_node->params.u.hdr.rmvParams.type =
+					e_FM_PCD_MANIP_RMV_GENERIC;
 		hm_node->params.u.hdr.rmvParams.u.generic.offset =
 					premove_hm->remove_params.custom.offset;
 		hm_node->params.u.hdr.rmvParams.u.generic.size =
@@ -5556,7 +5562,7 @@ static int vlan_hm_update_params(struct dpa_cls_hm *pvlan_hm)
 	switch (pvlan_hm->vlan_params.type) {
 	case DPA_CLS_HM_VLAN_INGRESS:
 		hm_node->params.u.hdr.rmv = TRUE;
-		hm_node->params.u.hdr.rmvParams.type	=
+		hm_node->params.u.hdr.rmvParams.type =
 					e_FM_PCD_MANIP_RMV_BY_HDR;
 		hm_node->params.u.hdr.rmvParams.u.byHdr.type =
 					e_FM_PCD_MANIP_RMV_BY_HDR_SPECIFIC_L2;
-- 
1.7.5.4

